types/camelize/index.d.ts

```diff
@@ -7,14 +7,14 @@ type CamelCase<S extends string> = S extends `${infer FirstWord}${CamelSeparator
     ? `${Lowercase<FirstWord>}${Uppercase<SecondChar>}${CamelCase<Remaining>}`
     : Lowercase<S>;
 
-type Camelize<T> = {
-    [K in keyof T as CamelCase<string & K>]: T[K] extends {} ? Camelize<T[K]> : T[K];
+export type Camelize<T> = {
+    [K in keyof T as CamelCase<string & K>]: T[K] extends Date ? T[K]
+        : T[K] extends RegExp ? T[K]
+        : T[K] extends Array<infer I> ? Array<Camelize<I>>
+        : T[K] extends {} ? Camelize<T[K]>
+        : T[K];
 };
 
-declare function camelize<T>(obj: T): T extends string ? CamelCase<T>
-    : T extends RegExp ? T
-    : T extends Date ? T
```

Comment:
check for `RegExp` and `Date` happens in the `walk` function which is also run on nested fields. Moving this condition to the `Camelize` type